File: tests/python/test_data_info.py
"""Tests for the read-only per-array data summary."""

import numpy as np
import pytest

import meshioplusplus as mp
from meshioplusplus._data_info import _info_py

from .helpers_data import data_mesh


def by_name(arrays, location, name):
    for a in arrays:
        if a["location"] == location and a["name"] == name:
            return a
    raise AssertionError(f"{location}:{name} not in the report")


def test_counts_every_array():
    arrays = mp.data_info(data_mesh())
    assert len(arrays) == 5
    locations = [a["location"] for a in arrays]
    assert locations.count("point_data") == 2
    assert locations.count("cell_data") == 2
    assert locations.count("field_data") == 1


def test_grouped_by_location_then_sorted_by_name():
    arrays = mp.data_info(data_mesh())
    assert [(a["location"], a["name"]) for a in arrays] == [
        ("point_data", "T"),
        ("point_data", "v"),
        ("cell_data", "mat"),
        ("cell_data", "tag"),
        ("field_data", "meta"),
    ]


def test_scalar_statistics_match_hand_computed_values():
    a = by_name(mp.data_info(data_mesh()), "point_data", "T")
    # T = {0, 1, 11, 10, 2, 12}
    assert a["num_entries"] == 6
    assert a["num_components"] == 1
    assert a["num_values"] == 6
    assert a["min"] == pytest.approx(0.0)
    assert a["max"] == pytest.approx(12.0)
    assert a["mean"] == pytest.approx(36.0 / 6.0)
    assert a["num_nan"] == 0
    assert a["num_inf"] == 0
    assert a["num_finite"] == 6


def test_vector_reports_shape_and_per_component_statistics():
    a = by_name(mp.data_info(data_mesh()), "point_data", "v")
    assert a["shape"] == (6, 3)
    assert a["num_components"] == 3
    assert a["num_entries"] == 6
    assert a["num_values"] == 18
    # Component 0 of v is {1,0,0,1,2,0}
    assert a["min_per_component"][0] == pytest.approx(0.0)
    assert a["max_per_component"][0] == pytest.approx(2.0)
    assert a["mean_per_component"][0] == pytest.approx(4.0 / 6.0)
    assert a["max_per_component"][2] == pytest.approx(1.0)


def test_cell_data_spans_every_block():
    a = by_name(mp.data_info(data_mesh()), "cell_data", "mat")
    assert a["num_blocks"] == 2
    assert a["num_entries"] == 3  # 2 triangles + 1 quad
    assert a["min"] == pytest.approx(1.0)
    assert a["max"] == pytest.approx(3.0)
    assert a["mean"] == pytest.approx(2.0)
    assert a["inconsistent_blocks"] is False


def test_dtype_reported_as_stored():
    arrays = mp.data_info(data_mesh())
    assert by_name(arrays, "point_data", "T")["dtype"] == "f8"
    assert by_name(arrays, "cell_data", "tag")["dtype"] == "i4"


def test_nan_and_inf_counts_are_correct():
    m = data_mesh()
    m.point_data["n"] = np.array([1.0, np.nan, np.inf, -np.inf, 2.0, np.nan])
    a = by_name(mp.data_info(m), "point_data", "n")
    assert a["num_nan"] == 2
    assert a["num_inf"] == 2
    assert a["num_finite"] == 2
    assert a["num_values"] == 6
    # Reductions ignore the non-finite values entirely.
    assert a["min"] == pytest.approx(1.0)
    assert a["max"] == pytest.approx(2.0)
    assert a["mean"] == pytest.approx(1.5)


def test_all_non_finite_array_yields_nan_statistics():
    m = data_mesh()
    m.point_data["n"] = np.full(6, np.nan)
    a = by_name(mp.data_info(m), "point_data", "n")
    assert a["num_finite"] == 0
    assert np.isnan(a["min"])
    assert np.isnan(a["max"])
    assert np.isnan(a["mean"])


def test_mesh_with_no_data():
    m = mp.Mesh(
        np.array([[0.0, 0, 0], [1, 0, 0], [0, 1, 0]]),
        [("triangle", np.array([[0, 1, 2]]))],
    )
    assert mp.data_info(m) == []


def test_does_not_modify_the_mesh():
    m = data_mesh()
    before = sorted(m.point_data)
    mp.data_info(m)
    assert sorted(m.point_data) == before


def test_cpp_matches_python():
    pytest.importorskip("meshioplusplus._core")
    m = data_mesh()
    m.point_data["n"] = np.array([1.0, np.nan, np.inf, -np.inf, 2.0, np.nan])
    cpp = mp.data_info(m)
    py = _info_py(m)
    assert len(cpp) == len(py)
    for a, b in zip(cpp, py):
        assert a["location"] == b["location"]
        assert a["name"] == b["name"]
        for key in (
            "num_blocks",
            "num_entries",
            "num_components",
            "num_values",
            "num_nan",
            "num_inf",
            "num_finite",
            "inconsistent_blocks",
        ):
            assert a[key] == b[key], (a["name"], key)
        for key in ("min", "max", "mean"):
            assert np.isclose(a[key], b[key], equal_nan=True), (a["name"], key)

File: tests/cpp/test_data_info.cpp
//  ██████   ██████ ██████████  █████████  █████   █████ █████    ███████
// ░░██████ ██████ ░░███░░░░░█ ███░░░░░███░░███   ░░███ ░░███   ███░░░░░███      ███         ███
//  ░███░█████░███  ░███  █ ░ ░███    ░░░  ░███    ░███  ░███  ███     ░░███    ░███        ░███
//  ░███░░███ ░███  ░██████   ░░█████████  ░███████████  ░███ ░███      ░███ ███████████ ███████████
//  ░███ ░░░  ░███  ░███░░█    ░░░░░░░░███ ░███░░░░░███  ░███ ░███      ░███░░░░░███░░░ ░░░░░███░░░
//  ░███      ░███  ░███ ░   █ ███    ░███ ░███    ░███  ░███ ░░███     ███     ░███        ░███
//  █████     █████ ██████████░░█████████  █████   █████ █████ ░░░███████░      ░░░         ░░░
// ░░░░░     ░░░░░ ░░░░░░░░░░  ░░░░░░░░░  ░░░░░   ░░░░░ ░░░░░    ░░░░░░░
//
//
//  License:         MIT License
//                   meshio++ default license: LICENSE
//
//  Main authors:    Vicente Mataix Ferrandiz
//
//
// Tests for the read-only per-array data summary.

// System includes
#include <cmath>
#include <limits>
#include <stdexcept>
#include <string>
#include <vector>

// External includes
#include <gtest/gtest.h>

// Project includes
#include "mesh_fixtures.hpp"
#include "meshioplusplus/operations/data_info.hpp"

namespace {

using meshioplusplus::data_array_info;
using meshioplusplus::data_info;
using meshioplusplus::DataArrayInfo;
using meshioplusplus::DataInfoReport;
using meshioplusplus::DataLocation;
using meshioplusplus::Mesh;

const DataArrayInfo* find(const DataInfoReport& rReport, DataLocation loc,
                          const std::string& rName) {
    for (const DataArrayInfo& a : rReport.mArrays)
        if (a.mLocation == loc && a.mName == rName)
            return &a;
    return nullptr;
}

TEST(DataInfo, CountsEveryArray) {
    DataInfoReport r = data_info(mt::data_mesh());
    EXPECT_EQ(r.mNumPointData, 2);  // T, v
    EXPECT_EQ(r.mNumCellData, 2);   // mat, tag
    EXPECT_EQ(r.mNumFieldData, 1);  // meta
    EXPECT_EQ(r.mArrays.size(), 5u);
}

TEST(DataInfo, ArraysAreGroupedByLocationThenSortedByName) {
    DataInfoReport r = data_info(mt::data_mesh());
    ASSERT_EQ(r.mArrays.size(), 5u);
    EXPECT_EQ(r.mArrays[0].mLocation, DataLocation::Point);
    EXPECT_EQ(r.mArrays[0].mName, "T");
    EXPECT_EQ(r.mArrays[1].mName, "v");
    EXPECT_EQ(r.mArrays[2].mLocation, DataLocation::Cell);
    EXPECT_EQ(r.mArrays[2].mName, "mat");
    EXPECT_EQ(r.mArrays[3].mName, "tag");
    EXPECT_EQ(r.mArrays[4].mLocation, DataLocation::Field);
    EXPECT_EQ(r.mArrays[4].mName, "meta");
}

TEST(DataInfo, ScalarPointArrayStatistics) {
    DataInfoReport r = data_info(mt::data_mesh());
    const DataArrayInfo* a = find(r, DataLocation::Point, "T");
    ASSERT_NE(a, nullptr);
    // T = {0, 1, 11, 10, 2, 12}
    EXPECT_EQ(a->mNumEntries, 6);
    EXPECT_EQ(a->mNumComponents, 1);
    EXPECT_EQ(a->mNumValues, 6);
    EXPECT_DOUBLE_EQ(a->mMin, 0.0);
    EXPECT_DOUBLE_EQ(a->mMax, 12.0);
    EXPECT_NEAR(a->mMean, 36.0 / 6.0, 1e-12);
    EXPECT_EQ(a->mNumNan, 0);
    EXPECT_EQ(a->mNumInf, 0);
    EXPECT_EQ(a->mNumFinite, 6);
}

TEST(DataInfo, VectorArrayReportsPerComponentStatistics) {
    DataInfoReport r = data_info(mt::data_mesh());
    const DataArrayInfo* a = find(r, DataLocation::Point, "v");
    ASSERT_NE(a, nullptr);
    EXPECT_EQ(a->mNumComponents, 3);
    EXPECT_EQ(a->mNumEntries, 6);
    EXPECT_EQ(a->mNumValues, 18);
    ASSERT_EQ(a->mMinPerComponent.size(), 3u);
    // Component 0 of v is {1,0,0,1,2,0}
    EXPECT_DOUBLE_EQ(a->mMinPerComponent[0], 0.0);
    EXPECT_DOUBLE_EQ(a->mMaxPerComponent[0], 2.0);
    EXPECT_NEAR(a->mMeanPerComponent[0], 4.0 / 6.0, 1e-12);
    // Component 2 is {0,0,1,0,0,0}
    EXPECT_DOUBLE_EQ(a->mMaxPerComponent[2], 1.0);
}

TEST(DataInfo, CellDataSpansEveryBlock) {
    DataInfoReport r = data_info(mt::data_mesh());
    const DataArrayInfo* a = find(r, DataLocation::Cell, "mat");
    ASSERT_NE(a, nullptr);
    EXPECT_EQ(a->mNumBlocks, 2);
    // 2 triangle cells + 1 quad cell
    EXPECT_EQ(a->mNumEntries, 3);
    EXPECT_EQ(a->mNumValues, 3);
    EXPECT_DOUBLE_EQ(a->mMin, 1.0);
    EXPECT_DOUBLE_EQ(a->mMax, 3.0);
    EXPECT_NEAR(a->mMean, 2.0, 1e-12);
    EXPECT_FALSE(a->mInconsistentBlocks);
}

TEST(DataInfo, ReportsDtypeAsStored) {
    DataInfoReport r = data_info(mt::data_mesh());
    const DataArrayInfo* t = find(r, DataLocation::Point, "T");
    ASSERT_NE(t, nullptr);
    EXPECT_TRUE(meshioplusplus::detail::is_float_dtype(t->mDtype));
    const DataArrayInfo* tag = find(r, DataLocation::Cell, "tag");
    ASSERT_NE(tag, nullptr);
    EXPECT_FALSE(meshioplusplus::detail::is_float_dtype(tag->mDtype));
}

TEST(DataInfo, CountsNaNAndInf) {
    Mesh m = mt::data_mesh();
    const double nan = std::numeric_limits<double>::quiet_NaN();
    const double inf = std::numeric_limits<double>::infinity();
    m.AddPointData("n", mt::data_array({1.0, nan, inf, -inf, 2.0, nan}));
    DataInfoReport r = data_info(m);
    const DataArrayInfo* a = find(r, DataLocation::Point, "n");
    ASSERT_NE(a, nullptr);
    EXPECT_EQ(a->mNumNan, 2);
    EXPECT_EQ(a->mNumInf, 2);
    EXPECT_EQ(a->mNumFinite, 2);
    EXPECT_EQ(a->mNumValues, 6);
    // Reductions ignore the non-finite values entirely.
    EXPECT_DOUBLE_EQ(a->mMin, 1.0);
    EXPECT_DOUBLE_EQ(a->mMax, 2.0);
    EXPECT_NEAR(a->mMean, 1.5, 1e-12);
}

TEST(DataInfo, AllNonFiniteArrayYieldsNaNStatistics) {
    Mesh m = mt::data_mesh();
    const double nan = std::numeric_limits<double>::quiet_NaN();
    m.AddPointData("n", mt::data_array({nan, nan, nan, nan, nan, nan}));
    const DataArrayInfo a = data_array_info(m, DataLocation::Point, "n");
    EXPECT_EQ(a.mNumFinite, 0);
    EXPECT_TRUE(std::isnan(a.mMin));
    EXPECT_TRUE(std::isnan(a.mMax));
    EXPECT_TRUE(std::isnan(a.mMean));
}

TEST(DataInfo, EmptyMeshHasNoArrays) {
    Mesh m;
    DataInfoReport r = data_info(m);
    EXPECT_EQ(r.mArrays.size(), 0u);
    EXPECT_EQ(r.mNumPointData, 0);
    EXPECT_EQ(r.mNumCellData, 0);
    EXPECT_EQ(r.mNumFieldData, 0);
}

TEST(DataInfo, MeshWithGeometryButNoDataHasNoArrays) {
    DataInfoReport r = data_info(mt::tri_mesh());
    EXPECT_EQ(r.mArrays.size(), 0u);
}

TEST(DataInfo, SingleArrayLookup) {
    const DataArrayInfo a = data_array_info(mt::data_mesh(), DataLocation::Field, "meta");
    EXPECT_EQ(a.mName, "meta");
    EXPECT_EQ(a.mNumEntries, 3);
    EXPECT_DOUBLE_EQ(a.mMin, 1.0);
    EXPECT_DOUBLE_EQ(a.mMax, 3.0);
}

TEST(DataInfo, UnknownNameThrows) {
    EXPECT_THROW(data_array_info(mt::data_mesh(), DataLocation::Point, "nope"),
                 std::invalid_argument);
}

TEST(DataInfo, DoesNotModifyTheMesh) {
    Mesh in = mt::data_mesh();
    const std::size_t before = in.NumPointData();
    data_info(in);
    EXPECT_EQ(in.NumPointData(), before);
    EXPECT_TRUE(in.HasPointData("T"));
}

}  // namespace
